Added materialized view and duplicated v2 Tinybird endpoints

cmraible opened the pull request:
ref https://linear.app/ghost/issue/NY-865/analytics-sources-not-populating-for-tangle-due-to-408-timeouts

## Problem

Tinybird endpoints are timing out in production for the sites with the most data. the `mv_session_data` pipe, which almost all our endpoints depend on, calculates complex aggregations at query time, which we believe to be the biggest contributor to poor performance when querying against large data sets.

## Fix

The solution to fix this is to convert `mv_session_data` from a pipe to a materialized view, such that Tinybird will calculate these aggregations as ingest time instead of at query time. As this is a rather large change to the implementation, we've opted to create a duplicate v2 pipeline rather than updating the existing pipeline in place. This way we can validate the v2 pipeline in production against real production data, without actually changing any of the user-facing behavior, before cutting over to using the v2 pipeline.

---

## Changes made

- Creates a new datasource for the materialized view for `mv_session_data`
- Creates mv_session_data_v2 pipe to feed the materialized view
- Creates filtered_sessions_v2, which is backed by the new materialized view
- Creates v2 of all endpoints, which are backed by the filtered_sessions_v2 and the new materialized view
- Duplicates all test files for endpoints, without making any changes to the test cases or expected results, to ensure that v1 and v2 of our endpoints maintain the same behavior
- Adds ability to switch Ghost to use v2 endpoints for the sake of validating against live production data

Creating a duplicate pipeline makes it easier to validate this in production, but it does make reviewing these changes more difficult, since the git diff doesn't clearly show what has changed in each endpoint. I've added a comment to each file that shows the diff of the v2 endpoint against the original unversioned endpoint to make reviewing this PR easier, but you'll have to run those commands locally with this branch checked out to see the changes.

## Testing Ghost against v2 endpoints

This [commit](https://github.com/TryGhost/Ghost/pull/25719/commits/87d6b679713cee0632590bd511023be04219f94d) adds the ability to point Ghost to the v2 version of the endpoints. To use this, set `tinybird:stats:version` to `v2` in your `config.local.json`, then re-run `yarn dev:analytics`.

cmraible committed:
Added materialized view for mv_session_data pipe

cmraible committed:
Updated e2e test to replace hard-coded waitForTimeout calls with a more robust polling approach

cmraible committed:
Updated top_utm_content to top_utm_contents to match filename

cmraible committed:
Reverted changes to v1 pipes

cmraible committed:
Created v2 pipes using materialized views

cmraible committed:
Added tests for all v2 endpoints

cmraible committed:
Added v2 endpoints and tests for top_locations and top_pages

cmraible committed:
Removed e2e test that will fail since frontend isn't using v2 endpoints

cmraible committed:
Reverted incidental change to mv_session_data.pipe

cmraible commented on ghost/core/core/server/data/tinybird/endpoints/api_top_pages_v2.pipe:
```diff
@@ -0,0 +1,39 @@
+TOKEN "stats_page" READ
```
```
diff -u ghost/core/core/server/data/tinybird/endpoints/api_top_pages.pipe ghost/core/core/server/data/tinybird/endpoints/api_top_pages_v2.pipe
```

cmraible commented on ghost/core/core/server/data/tinybird/endpoints/api_top_sources_v2.pipe:
```diff
@@ -0,0 +1,28 @@
+TOKEN "stats_page" READ
```
```
diff -u ghost/core/core/server/data/tinybird/endpoints/api_top_sources.pipe ghost/core/core/server/data/tinybird/endpoints/api_top_sources_v2.pipe
```

cmraible commented on ghost/core/core/server/data/tinybird/endpoints/api_top_utm_sources_v2.pipe:
```diff
@@ -0,0 +1,30 @@
+TOKEN "stats_page" READ
```
```
diff -u ghost/core/core/server/data/tinybird/endpoints/api_top_utm_sources.pipe ghost/core/core/server/data/tinybird/endpoints/api_top_utm_sources_v2.pipe
```

cmraible commented on ghost/core/core/server/data/tinybird/pipes/mv_session_data_v2.pipe:
```diff
@@ -0,0 +1,22 @@
+TOKEN "axis" READ
```
```
diff -u ghost/core/core/server/data/tinybird/pipes/mv_session_data.pipe ghost/core/core/server/data/tinybird/pipes/mv_session_data_v2.pipe
```

cmraible commented:
Working on a clean way to toggle the frontend to use the v2 version of the endpoints so we can test out the new pipeline on the frontend as well. We'll need this to help validate the v2 implementation against the existing live implementation.

troyciesco commented on ghost/core/core/server/data/tinybird/endpoints/api_kpis_v2.pipe:
```diff
@@ -0,0 +1,162 @@
+            {% end %}
+            sd.session_id,
+            pageviews,
+            pageviews = 1 as is_bounce,
```
this caught my eye at first and i was about to ask about it - but i see it comes from what was originally in `mv_session_data.pipe`

cmraible replied:
Yeah we could probably remove this because we don't show bounce rate anywhere in the UI anymore, but we can handle that separately 🙂

9larsons replied:
I believe John is actually using bounce rate in axis. We need to revisit Tinybird use outside of Ghost at some point soon.

cmraible committed:
Added ability to point Ghost to v2 of endpoints

cmraible commented on ghost/core/core/server/data/tinybird/endpoints/api_post_visitor_counts_v2.pipe:
```diff
@@ -0,0 +1,17 @@
+TOKEN "stats_page" READ
```
No changes in this pipe compared to the original, since this only uses _mv_hits. Created v2 for consistency so we can configure Ghost to use v2.

9larsons reviewed (changes requested):
Question about local use.

cmraible committed:
Removed exception for applying version when running with local enabled

9larsons reviewed (approved):
Looks good, looking forward to playing around and seeing the diff with real data!

cmraible commented:
Deployment check output to confirm this won't change any existing endpoints:

```

Running against Tinybird Cloud: Workspace web_analytics
Authentication method: token via --token option

» Validating deployment...


* Changes to be deployed:
------------------------------------------------
status: new
name: _mv_session_data_v2
type: datasource
path: datasources/_mv_session_data_v2.datasource
------------------------------------------------
status: new
name: api_top_sources_v2
type: endpoint
path: endpoints/api_top_sources_v2.pipe
------------------------------------------------
status: new
name: api_top_devices_v2
type: endpoint
path: endpoints/api_top_devices_v2.pipe
------------------------------------------------
status: new
name: api_kpis_v2
type: endpoint
path: endpoints/api_kpis_v2.pipe
------------------------------------------------
status: new
name: api_post_visitor_counts_v2
type: endpoint
path: endpoints/api_post_visitor_counts_v2.pipe
------------------------------------------------
status: new
name: api_top_utm_sources_v2
type: endpoint
path: endpoints/api_top_utm_sources_v2.pipe
------------------------------------------------
status: new
name: api_top_utm_contents_v2
type: endpoint
path: endpoints/api_top_utm_contents_v2.pipe
------------------------------------------------
status: new
name: api_top_utm_terms_v2
type: endpoint
path: endpoints/api_top_utm_terms_v2.pipe
------------------------------------------------
status: new
name: api_top_utm_mediums_v2
type: endpoint
path: endpoints/api_top_utm_mediums_v2.pipe
------------------------------------------------
status: new
name: api_active_visitors_v2
type: endpoint
path: endpoints/api_active_visitors_v2.pipe
------------------------------------------------
status: new
name: api_top_locations_v2
type: endpoint
path: endpoints/api_top_locations_v2.pipe
------------------------------------------------
status: new
name: api_top_utm_campaigns_v2
type: endpoint
path: endpoints/api_top_utm_campaigns_v2.pipe
------------------------------------------------
status: new
name: api_top_pages_v2
type: endpoint
path: endpoints/api_top_pages_v2.pipe
------------------------------------------------
status: new
name: mv_session_data_v2
type: materialization
path: pipes/mv_session_data_v2.pipe
------------------------------------------------
status: new
name: filtered_sessions_v2
type: pipe
path: pipes/filtered_sessions_v2.pipe
------------------------------------------------

* Changes in tokens to be deployed:
------------------------------------
Change: updated
Token name: axis
Added permissions:
api_active_visitors_v2.pipe:READ
api_kpis_v2.pipe:READ
api_post_visitor_counts_v2.pipe:READ
api_top_devices_v2.pipe:READ
api_top_locations_v2.pipe:READ
api_top_pages_v2.pipe:READ
api_top_sources_v2.pipe:READ
api_top_utm_campaigns_v2.pipe:READ
api_top_utm_contents_v2.pipe:READ
api_top_utm_mediums_v2.pipe:READ
api_top_utm_sources_v2.pipe:READ
api_top_utm_terms_v2.pipe:READ
filtered_sessions_v2.pipe:READ
mv_session_data_v2.pipe:READ
Removed permissions: 
------------------------------------
Change: updated
Token name: stats_page
Added permissions:
api_active_visitors_v2.pipe:READ
api_kpis_v2.pipe:READ
api_post_visitor_counts_v2.pipe:READ
api_top_devices_v2.pipe:READ
api_top_locations_v2.pipe:READ
api_top_pages_v2.pipe:READ
api_top_sources_v2.pipe:READ
api_top_utm_campaigns_v2.pipe:READ
api_top_utm_contents_v2.pipe:READ
api_top_utm_mediums_v2.pipe:READ
api_top_utm_sources_v2.pipe:READ
api_top_utm_terms_v2.pipe:READ
Removed permissions: 
------------------------------------

 Data that will be copied with this deployment:
--------------------------------------------------------------------
| Datasource          | Backfill type                              |
--------------------------------------------------------------------
| _mv_session_data_v2 | Using Materialized Pipe mv_session_data_v2 |
--------------------------------------------------------------------


✓ Deployment is valid
```

cmraible committed:
Removed tb_version from allowed options in topContent endpoint
